components/user-mgt/org.wso2.carbon.identity.user.pre.update.profile.action/src/main/java/org/wso2/carbon/identity/user/pre/update/profile/action/internal/execution/PreUpdateProfileRequestBuilder.java

```diff
@@ -0,0 +1,327 @@
+            throws ActionExecutionRequestBuilderException {
+
+        String userStoreDomain = userActionContext.getUserActionRequestDTO().getUserStoreDomain();
+        if (StringUtils.isBlank(userStoreDomain)) {
+
+            try {
+                org.wso2.carbon.user.core.common.User userFromUserStore =
+                        userStoreManager.getUserWithID(userActionContext.getUserActionRequestDTO().getUserId(), null,
+                                UserCoreConstants.DEFAULT_PROFILE);
+                userStoreDomain = userFromUserStore.getUserStoreDomain();
+            } catch (org.wso2.carbon.user.core.UserStoreException e) {
+                throw new ActionExecutionRequestBuilderException("Error while retrieving user store domain.", e);
+            }
+        }
+
+        return new UserStore(userStoreDomain);
+    }
+
+    private UniqueIDUserStoreManager getUserStoreManager() throws ActionExecutionRequestBuilderException {
+
+        String tenantDomain = IdentityContext.getThreadLocalIdentityContext().getTenantDomain();
+        RealmService realmService = PreUpdateProfileActionServiceComponentHolder.getInstance().getRealmService();
+
+        if (realmService == null) {
+            throw new ActionExecutionRequestBuilderException("Realm service is unavailable.");
+        }
+
+        try {
+            int tenantId = realmService.getTenantManager().getTenantId(tenantDomain);
+            UserRealm userRealm = realmService.getTenantUserRealm(tenantId);
+
+            if (userRealm == null) {
+                throw new ActionExecutionRequestBuilderException(
+                        "User realm is not available for tenant: " + tenantDomain);
+            }
+
+            UserStoreManager userStoreManager = userRealm.getUserStoreManager();
+            if (!(userStoreManager instanceof UniqueIDUserStoreManager)) {
+                throw new ActionExecutionRequestBuilderException(
+                        "User store manager is not an instance of UniqueIDUserStoreManager for tenant: " +
+                                tenantDomain);
+            }
+
+            return (UniqueIDUserStoreManager) userStoreManager;
+        } catch (UserStoreException e) {
+            throw new ActionExecutionRequestBuilderException(
+                    "Error while loading user store manager for tenant: " + tenantDomain, e);
+        }
+    }
+
+    private boolean isMultiValuedClaim(String claimUri) throws ActionExecutionRequestBuilderException {
+
+        ClaimMetadataManagementService claimMetadataManagementService =
+                PreUpdateProfileActionServiceComponentHolder.getInstance().getClaimManagementService();
+
+        try {
+            Optional<LocalClaim>
+                    localClaim = claimMetadataManagementService.getLocalClaim(claimUri,
+                    IdentityContext.getThreadLocalIdentityContext().getTenantDomain());
```

Comment:
Shall pass the tenant domain as a method paramter instead of resolve per claim.

Reply:
In this context this is the just the second reference for the tenant domain. So passing that through will overall clutter many other methods.
Moreover, tenant domain is overall not passed to this service as input. Thus proper functioning depends only on the value in IdentityContext, which is an expectation.

Reply:
Also the maximum number of time this loop can be executed if 10 as we allow only 10 claims to be sent to the extension